lib/src/responder/simple/provider.dart

```diff
@@ -0,0 +1,464 @@
+  /// When [addToTree] is false, the node will not be inserted into the node provider.
+  /// When [init] is false, onCreated() is not called.
+  LocalNode getOrCreateNode(String path, [bool addToTree = true, bool init = true]) {
+    LocalNode node = _getNode(path, allowStubs: true);
+
+    if (node != null) {
+      if (addToTree) {
+        Path po = new Path(path);
+        if (!po.isRoot) {
+          LocalNode parent = getNode(po.parentPath);
+
+          if (parent != null && !parent.children.containsKey(po.name)) {
+            parent.addChild(po.name, node);
+            parent.listChangeController.add(po.name);
+            node.listChangeController.add(r"$is");
+          }
+        }
+
+        if (node is SimpleNode) {
+          node._stub = false;
+        }
+      }
+
+      return node;
+    }
+
+    if (addToTree) {
+      return createNode(path, init);
+    } else {
+      node = new SimpleNode(path, this)
+        .._stub = true;
+      nodes[path] = node;
+      return node;
+    }
+  }
+
+  /// Checks if this provider has the node at [path].
+  bool hasNode(String path) {
+    SimpleNode node = nodes[path];
+
+    if (node == null) {
+      return false;
+    }
+
+    if (node.isStubNode == true) {
+      return false;
+    }
+
+    return true;
+  }
+
+  void registerResolver(SimpleNodeFactory factory) {
+    if (!_resolverFactories.contains(factory)) {
+      _resolverFactories.add(factory);
+    }
+  }
+
+  void unregisterResolver(SimpleNodeFactory factory) {
+    _resolverFactories.remove(factory);
+  }
+
+  @override
+  void addProfile(String name, NodeFactory factory) {
+    _profiles[name] = factory;
+  }
+
+  /// Sets the function that persists the nodes.
+  void setPersistFunction(ExecutableFunction doPersist) {
+    _persist = doPersist;
+  }
+
+  /// Persist the nodes in this provider.
+  /// If you are not using a LinkProvider, then call [setPersistFunction] to set
+  /// the function that is called to persist.
+  void persist([bool now = false]) {
+    if (now) {
+      if (_persist == null) {
+        return;
+      }
+
+      _persist();
+    } else {
+      new Future.delayed(const Duration(seconds: 5), () {
+        if (_persist == null) {
+          return;
+        }
+
+        _persist();
+      });
+    }
+  }
+
+  /// Creates a node at [path].
+  /// If a node already exists at this path, an exception is thrown.
+  /// If [init] is false, onCreated() is not called.
+  SimpleNode createNode(String path, [bool init = true]) {
+    Path p = new Path(path);
+    LocalNode existing = nodes[path];
+
+    if (existing != null) {
+      if (existing is SimpleNode) {
+        if (existing._stub != true) {
+          throw new Exception("Node at ${path} already exists.");
+        } else {
+          existing._stub = false;
+        }
+      } else {
+        throw new Exception("Node at ${path} already exists.");
+      }
+    }
+
+    SimpleNode node = existing == null ? new SimpleNode(path, this) : existing;
```

Comment:
`existing ?? new SimpleNode(..)`

Reply:
it still need to work with dart 1.17

Reply:
@rinick Null aware operators have been in Dart since 1.12 so Joel's suggestion will work for 1.17